### PR TITLE
discovery: Prevent that SetListener gets called twice or after a tablet was added.

### PR DESCRIPTION
**Pull request by michael-berlin**
@guoliang100 

discovery: Fix datarace for "listener" field in HealthCheckImpl.

B=28616156

---
This change is [Reviewable](https://reviewable.io/reviews/youtube/vitess/1724)

### CODE REVIEW

File: go/vt/discovery/healthcheck.go [hunk 1]
```diff
@@ -745,7 +760,6 @@ func (hc *HealthCheckImpl) Close() error {
 	hc.mu.Lock()
 	defer hc.mu.Unlock()
 	close(hc.closeChan)
-	hc.listener = nil
 	for _, hcc := range hc.addrToConns {
```
Comment (guoliang100) [inline go/vt/discovery/healthcheck.go]: Is there any reason you do not want to reset listener?

**Comment by michael-berlin** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[go/vt/discovery/healthcheck.go, line 763 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1724#-KIUZHRJKc0Il7jggcoU-r1-763:-KIV4hJ4RLrV1r7zLAY8:2147263947) ([raw file](https://github.com/youtube/vitess/blob/9a9ef8375271058de81031ef045fdbe785dc1988/go/vt/discovery/healthcheck.go#L763)):*
<details><summary><i>Previously, guoliang100 (Liang) wrote…</i></summary>
> Is there any reason you do not want to reset listener?
</details>
How would this make a difference?

I assumed that a HealthCheck instance where Close() was called cannot be reused from there on. Therefore, it shouldn't make a difference if we clear the listener or not?

The only difference would be that we give up the reference on the listener and this way the GC can delete it if there aren't any other references? But I would assume that we give up on the HealthCheck instance eventually anyway when we called Close().

Let me know if I'm missing anything here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1724)*

**Comment by michael-berlin** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[go/vt/discovery/healthcheck.go, line 763 \[r1\]](https://reviewable.io:443/reviews/youtube/vitess/1724#-KIUZHRJKc0Il7jggcoU-r1-763:-KIV5cbvgzbE49EqJFjv:-2132187553) ([raw file](https://github.com/youtube/vitess/blob/9a9ef8375271058de81031ef045fdbe785dc1988/go/vt/discovery/healthcheck.go#L763)):*
<details><summary><i>Previously, michael-berlin (Michael Berlin) wrote…</i></summary>
> How would this make a difference?
> 
> I assumed that a HealthCheck instance where Close() was called cannot be reused from there on. Therefore, it shouldn't make a difference if we clear the listener or not?
> 
> The only difference would be that we give up the reference on the listener and this way the GC can delete it if there aren't any other references? But I would assume that we give up on the HealthCheck instance eventually anyway when we called Close().
> 
> Let me know if I'm missing anything here.
</details>
Just to clarify if this wasn't explicit to you: Not resetting the listener fixes the data race. That's why I put the fix in a separate commit :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/1724)*

**Comment by michael-berlin** [comment]
I'm merging this first and then https://github.com/youtube/vitess/pull/1726 which also includes the two commits from here.

**Pull request merged**